RPIPs/RPIP-43.md

```diff
@@ -8,37 +8,95 @@ status: Draft
 type: Protocol
 category: Core
 created: 2024-03-05
-requires: eip-4788
+requires: 30, eip-4788
 tags: tokenomics-2024, tokenomics-content
 ---
 
 ## Abstract
 This proposal drastically reduces the gas to add validators and distribute rewards for them. It does so by creating a single contract that can be used as the Ethereum withdrawal address for any number of validators. In addition, the contract will facilitate the application of node-level penalties, as opposed to the current state where penalties are assigned and processed at the minipool level. Note that this proposal also enables ETH-only Node operation.
 
 ## Specification
-- A Node Operator SHALL be able to deploy a single megapool contract
+
+### Megapool Contracts
+
+This specification introduces a type of smart contract called a "megapool" that
+handles Beacon Chain withdrawals for Rocket Pool validators deposited after
+this RPIP is implemented.
+
+- A Node Operator SHALL be able to deploy at most one megapool contract
     - This MAY be combined with the deposit of their first validator
-- A megapool contract SHALL use a proxy pattern that allows its functionality to be upgraded; the proxy implementation contract SHALL be called a megapool delegate
-- A Node Operator SHALL be able to add a validator to their existing megapool contract
-- A Node Operator SHALL be able to remove one or more exited validators in their megapool contract
+- Megapool contracts MUST be upgradable using a proxy pattern; a megapool's
+  proxy implementation contract SHALL be called the megapool's "delegate"
+- Legacy minipool deposits (ie, deposits of validators with withdrawal
+  credentials that are not a megapool) SHALL be disabled
+    - This SHOULD be achieved in a way that prevents creation of minipools
+
+### Validator Management
+
+Megapools keep track of the set of validators associated with them and the
+status of each validator. The possible statuses for validators associated with
+megapools SHALL include at least the following:
+  - Deposited: a Node Operator has added this validator to their megapool
+  - Dissolved: this validator's withdrawal credentials have been discovered to
+               be incorrect (ie, not this megapool)
```

Comment:
There's something off here... These states aren't _validator_ states. That should be beacon chain states.

I think we really need the ability to access both a beacon_chain_status (I suspect this has a proper name) and an rp_status associated. Eg, we need to know if a validator is exited on beacon chain (a beacon root thing) as well as if it was properly staking prior to that on RP. 

I think on the RP side, we could simplify to "pending_addition" (ie, we've done a deposit but haven't been matched), "in_megapool", "pending_removal" (a more action-oriented version of dissolved; could also be used for exited, but I think we need to refer to beacon root anyhow so I don't think that's helpful).

Reply:
I was intending these two states to be RP states, defined within this RPIP (just to be clear about that part at least).

Reply:
I'm still kinda confused on what you want changed here if anything.

Reply:
Right now the wording reads as a status about the actual validator, not an RP layer. I also proposed 3 different states that I think are more effective (both in use and in conveying meaning) than the suggested two.

Reply:
I'm fine with changing it, I'm just trying to figure out what to change it to. I'll draft something now and commit...

Minor question: why are your states lowercase snake case here, whereas in other places (e.g. Unstaking, Staking for megapool staked RPL) we use capitalised camel case for Rocket Pool states?